### DISCUSSION TITLE
Custom command to open pdf and note in a combined action

### DISCUSSION THREAD

**Discussion by tuurep** [papis/papis#920] [Q&A] [answered]
Hey I'm trying to dip my toes into writing a custom script for the first time reading references like
* https://papis.readthedocs.io/en/latest/scripting.html
* examples in [`examples/scripts`](https://github.com/papis/papis/tree/main/examples/scripts)

Looking for some gentle advice if anyone's feeling helpful:
* since I'm basically trying to do 2 existing commands on a document, is there some really easy way to combine them in a script?
* if not, would my best bet be to seek the source code for `papis open` and `papis edit -n` and try to merge them together as a new script?

I have a working helloworld as follows:

`~/.config/papis/scripts/papis-note`

```py
#!/usr/bin/env python3
# papis-short-help: Open document's note and pdf in a combined action

print("Hello")
```

**Comment by tuurep** [comment]
Looking at this now,

my current idea is to script it in bash, and use `papis list <flags?>`, then either:
- Read lines, open files ending in `.pdf` with `xdg-open`, and files ending in `.md` (..or notes extension of choice) with `$EDITOR`
- Parse lines ending in `.pdf` to one batch of arguments, `.md` to another, feed first batch to `<pdf viewer>`, second to `$EDITOR`

### Questions/problems:

#### Multiple flags for `papis list`

This would be very simple if I could do `papis list --file --notes` and it would list all `pdf`s and `md`s (only), but it will only do the first flag and ignore the other.

#### Custom scripting using python, purpose of `examples/scripts`

~~I was going to try out `papis.api` and what I could achieve with that, but doing `import papis.api` gives `ModuleNotFoundError: No module named 'papis'`~~

**Update:** didn't work when using `pipx` to install papis

Is it the intention that any script from `examples/scripts` could be moved to `.config/papis/scripts` and it would work?

**Update:** yes, works when using AUR package, tested with `examples/scripts/papis-check`

**Reply by tuurep** [reply]
Seems reasonable:

`papis list` (no flags) gets the document(s) dir(s)

Glob them like so:

`"$doc_dir"/*.pdf`
`"$doc_dir"/*.md`

First iteration:

```
#!/usr/bin/env bash
# papis-short-help: Open document's note and pdf in a combined action

query=$1

papis list "$query" | while read -r dir; do
        zathura "$dir"/*.pdf & disown
        $EDITOR "$dir"/*.md
done
```

Sorta works, with a couple things to fix:
1. With an empty query, does `papis list ""` which returns no documents
    * if query is empty, don't even insert an empty string
3. Replace `zathura` with `xdg-open` to be usable regardless of pdf viewer of choice
    * awkwardness: `xdg-open` can't take multiple files as args, write a wrapper func

Thought:
* When picking multiple documents, neovim will block until the next file and note can be opened. However this is consistent with papis in general, for example batch-opening note files with `papis edit -n`, they'll be opened one at a time instead of opening multiple buffers in neovim.
    * (therefore, I don't really care about this)

**update:** scrapped the above script in favor of python script [below](https://github.com/papis/papis/discussions/920#discussioncomment-10859550), handles all these edge cases much better

**Reply by alexfikl** [reply, marked as answer]
> `papis list` (no flags) gets the document(s) dir(s)

If you're going for a bash script. This is probably the easiest way to do it.

> Custom scripting using python, purpose of examples/scripts

I usually try to write my scripts in Python because it has easier command line parsing / validation and you can reuse some functionality that's not exactly exposed through the command line. 

The `papis-tags` script is a nice little example of this. For what you mentioned, some untested code would look like
```python
import click

import papis.api
import papis.cli


@click.command()
@papis.cli.query_argument()
def main(query: str) -> None:
    documents = papis.api.get_documents_in_lib(
        papis.api.get_lib_name(),
        search=query
    )
    documents = papis.api.pick_docs(documents)

    from papis.commands.edit import edit_notes
    from papis.commands.open import run as open_file

    for document in documents:
        open_file(document)
        edit_notes(document)

if __name__ == "__main__":
	main()
```
You can also use `document.get_files()` and `papis.api.open_file` to further filter what files you'd like to open. By default `open_file` above will open a picker to ask you to select the files you'd like, so that could be good enough.

**Reply by tuurep** [reply]
Ah, cool thanks!

I'll look at this later today but was just now wrapping up my thoughts so I can take a little break :smile:

**Reply by alexfikl** [reply]
Sounds good! Let me know if I can help some more (completely missed this question the first time).

By the way, for external scripts, we also export some environment variables that can be useful:
https://github.com/papis/papis/blob/6137701eeb7108e46afc2896936888aa4927feee/papis/commands/external.py#L36-L52

**Reply by tuurep** [reply]
> For what you mentioned, some untested code would look like...

Yes this works great (only `pick_docs` => `pick_doc`), thanks very much

I was wondering what would happen if a document has multiple files:
* with my bash script it opens all at once
* with your py script it opens another picker to choose which file to pick
   * can also open many at once using fzf multi pick

It's small things like these that make the python script feel more coherent when it reuses the existing funcs, so I think I won't fix the bash script any further :smile:

It's a bit of an issue though that I wouldn't really have figured out how to do it without your help the python way. Maybe in https://papis.readthedocs.io/en/latest/scripting.html there should be a python example in addition to the bash one? The one that you've cooked up here feels like it could be generally useful maybe?

> By the way, for external scripts, we also export some environment variables that can be useful:

Oh yeah also, I was looking for these earlier since the docs say:

> Papis defines environment variables such as `PAPIS_LIB` so that external scripts can make use of the user input.

But it feels like the rest is missing :D Could we add all the env variables with descriptions to the docs (maybe even in the same section)?

Also, I've occasionally seen some typos in the docs that I could easily fix just like that. So I've been thinking, I'd be interested in doing a full pass through the docs and fixing (just obvious) typos and do a PR. Should I go ahead?

**Reply by tuurep** [reply]
Ah I see https://papis.readthedocs.io/en/latest/api.html has good documentation of the API though,

but I was really yearning for some example that gives a general structure for something like:
1. Open a picker (with a query or not)
2. Pick document(s)
3. Do something with the results

**Reply by alexfikl** [reply]
> It's a bit of an issue though that I wouldn't really have figured out how to do it without your help the python way. Maybe in [papis.readthedocs.io/en/latest/scripting.html](https://papis.readthedocs.io/en/latest/scripting.html) there should be a python example in addition to the bash one? The one that you've cooked up here feels like it could be generally useful maybe?

Yeah, it definitely would help to add more examples like this. The plugin parts aren't generally very well documented :(
 
> Oh yeah also, I was looking for these earlier since the docs say:
> 
> > Papis defines environment variables such as `PAPIS_LIB` so that external scripts can make use of the user input.
> 
> But it feels like the rest is missing :D Could we add all the env variables with descriptions to the docs (maybe even in the same section)?

Yes! I don't recall when we expanded those, but apparently we forgot to update the docs :(

> Also, I've occasionally seen some typos in the docs that I could easily fix just like that. So I've been thinking, I'd be interested in doing a full pass through the docs and fixing (just obvious) typos and do a PR. Should I go ahead?

It's very welcome if you have the time! Multiple smaller PRs, as you find typos and badly documented parts, work too.

If you're feeling adventurous you could look into integrating something like https://github.com/crate-ci/typos into the CI. I imagine we'll have quite a few false positives due to all the URLs and whatnot, but it should catch some nice errors too :grin:

**Reply by tuurep** [reply]
Great, I'll for sure send a typofix PR in the near future

I may be interested in contributing the other (2) doc additions as well, but if you have a clear idea about them, please go ahead :D

My immediate thoughts about those:
* env variables could be listed in https://papis.readthedocs.io/en/latest/scripting.html under heading `Environment variables`
    * when referred to in docs, link to that anchor
* general `papis.api` picker structure either:
    1. divide https://papis.readthedocs.io/en/latest/scripting.html into `bash` and `python` sections
    2. or add an `Examples` section at the bottom of https://papis.readthedocs.io/en/latest/api.html with something like:
       ```py
        #!/usr/bin/env python3
        # papis-short-help: Help message for custom command

        import papis.api
        import papis.cli
        import click

        @click.command()
        @papis.cli.query_argument()
        def main(query: str) -> None:
            documents = papis.api.get_documents_in_lib(
                papis.api.get_lib_name(),
                search=query
            )
            documents = papis.api.pick_doc(documents)

            for document in documents:
                # Do something with document

        if __name__ == "__main__":
            main()

       ```

Typochecking in CI seems intimidating, if it nags on false positives would be so annoying :laughing: But interesting, didn't know about that.

**Reply by alexfikl** [reply]
> I may be interested in contributing the other (2) doc additions as well, but if you have a clear idea about them, please go ahead :D

That would be great! I haven't looked too much through that section of the docs, so I can't say I have a much better view of what would be nice to have :-?
 
> My immediate thoughts about those:
> 
> * env variables could be listed in [papis.readthedocs.io/en/latest/scripting.html](https://papis.readthedocs.io/en/latest/scripting.html) under heading `Environment variables`

I'm inclined to put these in a more prominent section. Maybe after the configuration file info? Those usually go together. Plus, some of those environment variables are also used outside of scripting, e.g. `PAPIS_DEBUG`.

> 1. divide [papis.readthedocs.io/en/latest/scripting.html](https://papis.readthedocs.io/en/latest/scripting.html) into `bash` and `python` sections
> 2. or add an `Examples` section at the bottom of [papis.readthedocs.io/en/latest/api.html](https://papis.readthedocs.io/en/latest/api.html) with something like:

These sound good to me! 

I would suggest to add the bigger chunks of code into the `examples` folder and then use `.. literalinclude:: ../../examples/scripts/script-name`. That way people can pick them up and we can call `flake8` and `mypy` and whatnot on them to make sure they stay up to date.

> Typochecking in CI seems intimidating, if it nags on false positives would be so annoying 😆 But interesting, didn't know about that.

Haha, yeah, but hopefully that's just an initial cost until we add all the useless stuff to some list of ignores. I've used typos on some projects and it seems pretty laid back. Plus it catches nice coding typos in variable names like `ExtensionMananger`.

**Reply by tuurep** [reply]
Thanks, all of that helps with getting some PRs started. I have something more pressing to get done right now, but might start later this week/weekend :)

**Reply by tuurep** [reply]
I've been working on doc PRs, but started having some hardware problems with my laptop 💀

Will pick it up again after getting that taken care of

**Reply by tuurep** [reply]
My laptop's fixed! I'm going ahead with the doc PRs :)
